### PR TITLE
[WIP] Changes to MessageEvent

### PR DESCRIPTION
**Pull request by windy1**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1114) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/548) | [Vanilla](https://github.com/SpongePowered/SpongeVanilla/pull/229) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/559)

This PR makes some substantial changes to how MessageEvents are handled. Instead of passing around a `Text` object, the event now contains a `MessageFormatter` which can be used to modify the event's message in a more reliable and modular fashion. This will make interactions and compatibility between chat plugins a lot easier. Previous methods like `getOriginalMessage`, `getRawMessage`, `getMessage`, and `setMessage` still behave as expected except instead of Optionals, a non-null Text is always returned. Developers should use `setMessageCancelled` to cancel the sending instead of setting the message to null.

----

Some example uses:

Let's say I would like to insert a "group" prefix to the message but I also would like to keep existing prefixes that other plugins may or may not have already applied. I can do something like:

```java
class GroupApplier extends SimpleTextTemplateApplier {
    public GroupApplier() {
        super(of('[', arg("group"), "] "));
    }
}
```

```java
@Listener
void onChat(MessageChannelEvent.Chat event) {
    GroupApplier applier = new GroupApplier();
    applier.setParameter("group", Text.builder("admin").color(TextColors.RED));
    event.getFormatter().getHeader().insert(0, applier);
}
```

On an unaltered message, this will produce something equivalent to:

`[<red>admin</red>] <NAME> MESSAGE`

Now lets say another plugin wants to override that prefix and make it say "mod" instead of "admin". Well then they can do:

```java
event.getHeader().forEach(GroupApplier.class, applier -> applier.setParameter("group", Text.builder("mod").color(TextColors.AQUA));
```

These are only a few of the many uses this PR will provide.

**Note:** This PR depends on https://github.com/SpongePowered/SpongeAPI/pull/1079. https://github.com/SpongePowered/SpongeAPI/pull/1079 should be merged into master *and then into bleeding* before this PR is merged.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java [hunk 1]
```diff
@@ -24,37 +24,254 @@
...
+
+    /**
+     * Parameter for body section in body partition.
+     */
+    String PARAM_MESSAGE_BODY = "body";
+
+    /**
+     * Parameter for footer section in footer partition.
+     */
+    String PARAM_MESSAGE_FOOTER = "footer";
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: "footer" doesn't make sense here, "suffix" does

https://github.com/SpongePowered/SpongeAPI/pull/1112#discussion_r54342362

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java [hunk 2]
```diff
@@ -24,37 +24,254 @@
...
 /**
  * Describes events when a involving a {@link Text} message.
  */
+@ImplementedBy(AbstractMessageEvent.class)
 public interface MessageEvent extends Event {
 
     /**
+     * Parameter for header section in header partition.
+     */
+    String PARAM_MESSAGE_HEADER = "header";
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: "header" doesn't make sense here, "prefix" does

https://github.com/SpongePowered/SpongeAPI/pull/1112#discussion_r54342363

File: src/main/java/org/spongepowered/api/text/transform/TextFormatter.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+     */
+    boolean add(E element);
+
+    /**
+     * Adds the specified elements to the end of this formatter.
+     *
+     * @param elements Elements to add
+     * @return True if the formatter changed as a result of the call
+     */
+    boolean add(Collection<E> elements);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/transform/TextFormatter.java]: Is there a reason for having this method when there's one that accepts `Iterable`? If so, can this (and related) be `default`ed?

File: src/main/java/org/spongepowered/api/text/transform/TextFormatter.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    /**
+     * Replaces the element previously at the specified index with the
+     * specified element.
+     *
+     * @param i Index to replace
+     * @param e Element to replace with
+     * @return Element previously at index
+     */
+    E set(int i, E e);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/transform/TextFormatter.java]: One char parameters aren't very friendly in my opinion - `index` and `element`

File: src/main/java/org/spongepowered/api/text/transform/SimpleTextTemplateApplier.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public TextTemplate getTemplate() {
+        return this.template;
+    }
+
+    @Override
+    public void setTemplate(TextTemplate template) {
+        checkNotNull(template, "template");
+        this.template = template;
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/transform/SimpleTextTemplateApplier.java]: Can do: `this.template = checkNotNull(template, "template");` (and for other cases like this)

File: src/main/java/org/spongepowered/api/text/transform/TextFormatter.java (hunk 3 above)
Comment (windy1) [inline src/main/java/org/spongepowered/api/text/transform/TextFormatter.java]: So we don't have to create a new Collection if we don't have to. I could probably add some defaults here.

File: src/main/java/org/spongepowered/api/text/sink/package-info.java [hunk 6]
```diff
@@ -22,7 +22,4 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/sink/package-info.java]: This package-info (api/text/sink/package-info.java) doesn't exist anymore, where did it come from? In any case, shouldn't be making any changes

File: src/main/java/org/spongepowered/api/text/transform/TextFormatter.java (hunk 3 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/transform/TextFormatter.java]: What do you mean create a new `Collection`? `Collection` extends `Iterable`

Comment (windy1) [inline src/main/java/org/spongepowered/api/text/transform/TextFormatter.java]: `List.addAll` takes a `Collection` so if we have an `Iterable` we need to create a new `Collection from it i.e `Lists.newArrayList(iterable)`,

Comment (kashike) [inline src/main/java/org/spongepowered/api/text/transform/TextFormatter.java]: http://docs.guava-libraries.googlecode.com/git-history/v17.0/javadoc/com/google/common/collect/Iterables.html#addAll(java.util.Collection, java.lang.Iterable)

File: src/main/java/org/spongepowered/api/event/achievement/GrantAchievementEvent.java [hunk 7]
```diff
@@ -41,6 +43,7 @@
      */
     Achievement getAchievement();
 
+    @ImplementedBy(AbstractMessageEvent.class)
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/event/achievement/GrantAchievementEvent.java]: This shouldn't be needed - just bump `event-gen-core` and `event-impl-gen` to `1.1.0` in `build.gradle`.

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java [hunk 8]
```diff
@@ -24,37 +24,254 @@
...
+    @PropertySettings(requiredParameter = false)
+    Text getOriginalMessage();
+
+    /**
+     * Returns true if the {@link Text} returned by
+     * {@link #getMessage()} should not be sent.
+     *
+     * @return True if message should not be sent
+     */
+    boolean isMessageCancelled();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: Shouldn't the event just be `Cancellable` instead?

Comment (windy1) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: There are events where you might want to cancel just the message and not the entire event itself. i.e. `KickPlayerEvent`

File: src/main/java/org/spongepowered/api/text/transform/FixedPartitionedTextFormatter.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.transform;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collection;
+<<<<<<< HEAD
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/text/transform/FixedPartitionedTextFormatter.java]: Merge conflict.

Comment (windy1) [inline src/main/java/org/spongepowered/api/text/transform/FixedPartitionedTextFormatter.java]: whoops, thanks

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java (hunk 2 above)
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: I personally agree with @kashike. Header/footer sounds like these are additional lines or something, not simply a prefix/suffix.

Comment (windy1) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: I still disagree tbh. `[mod][foo][bar]` looks like 3 prefixes, no? Well if we change the name then all of them fall under the single "prefix" category.

Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: Not a big problem, IMO.

**Pull request closed without merging**